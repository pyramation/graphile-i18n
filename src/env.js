import { cleanEnv, makeValidator } from 'envalid';

const array = makeValidator((x) => x.split(','), '');

export default cleanEnv(
  process.env,
  {
    ACCEPTED_LANGUAGES: array({ default: 'en,es' })
  },
  { dotEnvPath: null }
);
