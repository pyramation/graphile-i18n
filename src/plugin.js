const escapeIdentifier = (str) => `"${str.replace(/"/g, '""')}"`;
export const LangPlugin = (builder, options) => {
  const {
    langPluginLanguageCodeColumn = 'lang_code',
    langPluginLanguageCodeGqlField = 'langCode',
    langPluginAllowedTypes = ['citext', 'text'],
    langPluginDefaultLanguages = ['en']
  } = options;

  builder.hook('build', (build) => {
    const { pgSql: sql } = build;
    /** @type {import('graphile-build-pg').PgIntrospectionResultsByKind} */
    const introspection = build.pgIntrospectionResultsByKind;
    const inflection = build.inflection;

    const tablesWithLanguageTables = introspection.class.filter((table) =>
      table.tags.hasOwnProperty('i18n')
    );
    const tablesWithLanguageTablesIdInfo = introspection.class
      .filter((table) => table.tags.hasOwnProperty('i18n'))
      .map((table) => {
        return {
          identifier: table.primaryKeyConstraint.keyAttributes[0].name,
          idType: table.primaryKeyConstraint.keyAttributes[0].type.name
        };
      });
    const languageVariationTables = tablesWithLanguageTables.map((table) =>
      introspection.class.find(
        (t) =>
          t.name === table.tags.i18n && t.namespaceName === table.namespaceName
      )
    );

    const i18nTables = {};
    const tables = {};
    tablesWithLanguageTables.forEach((table, i) => {
      i18nTables[table.tags.i18n] = {
        table: table.name,
        key: null, // action_id
        connection: inflection.connection(inflection.tableType(table)),
        attrs: {},
        fields: {},
        keyInfo: tablesWithLanguageTablesIdInfo[i]
      };
      tables[table.name] = table.tags.i18n;
    });

    languageVariationTables.forEach((table) => {
      const foreignConstraintsThatMatter = table.constraints
        .filter((c) => c.type === 'f')
        .filter((c) => c.foreignClass.name === i18nTables[table.name].table);

      if (foreignConstraintsThatMatter.length !== 1)
        throw new Error(
          'lang table only supports one foreign key to parent table'
        );

      if (foreignConstraintsThatMatter[0].keyAttributes.length !== 1)
        throw new Error(
          'lang table only supports one non compound foreign key to parent table'
        );

      i18nTables[table.name].key =
        foreignConstraintsThatMatter[0].keyAttributes[0].name;

      const { identifier, idType } = i18nTables[table.name].keyInfo;

      table.attributes.forEach((attr) => {
        if ([langPluginLanguageCodeColumn, identifier].includes(attr.name))
          return;
        if (langPluginAllowedTypes.includes(attr.type.name)) {
          i18nTables[table.name].fields[inflection.column(attr)] = {
            type: attr.type.name,
            attr: attr.name,
            isNotNull: attr.isNotNull,
            column: inflection.column(attr)
          };
          i18nTables[table.name].attrs[attr.name] = {
            type: attr.type.name,
            attr: attr.name,
            column: inflection.column(attr)
          };
        }
      });
    });
    return build.extend(build, { i18n: { i18nTables, tables } });
  });

  builder.hook('GraphQLObjectType:fields', (fields, build, context) => {
    const {
      graphql: { GraphQLString, GraphQLObjectType, GraphQLNonNull },
      i18n: { i18nTables, tables },
      pgSql: sql
    } = build;
    const {
      scope: { pgIntrospection: table, isPgRowType },
      fieldWithHooks
    } = context;
    if (!isPgRowType || !table || table.kind !== 'class') {
      return fields;
    }

    const variationsTableName = tables[table.name];
    if (!variationsTableName) {
      return fields;
    }

    const i18nTable = i18nTables[variationsTableName];
    const { identifier, idType } = i18nTable.keyInfo;

    const { key, connection, attrs, fields: i18nFields } = i18nTable;

    const localeFieldName = 'localeStrings';
    const localeFieldsType = new GraphQLObjectType({
      name: `${context.Self.name}LocaleStrings`,
      description: `Locales for ${context.Self.name}`,
      fields: Object.keys(i18nFields).reduce(
        (memo, field) => {
          memo[field] = {
            type: i18nFields[field].isNotNull
              ? new GraphQLNonNull(GraphQLString)
              : GraphQLString,
            description: `Locale for ${field}`
          };
          return memo;
        },
        {
          langCode: {
            type: GraphQLString // MUST BE NULLABLE
          }
        }
      )
    });

    return build.extend(fields, {
      [localeFieldName]: fieldWithHooks(
        localeFieldName,
        (fieldContext) => {
          const { addDataGenerator } = fieldContext;
          addDataGenerator((parsedResolveInfoFragment) => {
            return {
              pgQuery: (queryBuilder) => {
                queryBuilder.select(
                  sql.fragment`${queryBuilder.getTableAlias()}.${sql.identifier(
                    identifier
                  )}`,
                  identifier
                );
              }
            };
          });
          const coalescedFields = Object.keys(i18nFields).map((field) => {
            const columnName = i18nFields[field].attr;
            const escColumnName = escapeIdentifier(columnName);
            const escFieldName = escapeIdentifier(field);
            return `coalesce(v.${escColumnName}, b.${escColumnName}) as ${escFieldName}`;
          });
          const props = {
            table,
            coalescedFields,
            variationsTableName,
            key
          };
          return {
            description: `Locales for ${context.Self.name}`,
            type: new GraphQLNonNull(localeFieldsType),
            async resolve({ id }, args, context) {
              const languageCodes =
                context.langCodes ?? langPluginDefaultLanguages;
              const dataloader = context.getLanguageDataLoader(
                props,
                context.pgClient,
                languageCodes,
                identifier,
                idType,
                langPluginLanguageCodeColumn,
                langPluginLanguageCodeGqlField
              );
              return dataloader.load(id);
            }
          };
        },
        'Adding the language code field from the lang plugin'
      )
    });
  });
};
