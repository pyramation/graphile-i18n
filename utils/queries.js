import { gql } from 'graphile-test';

export const GetProjectsAndLanguages = gql`
  query GetProjectsAndLanguagesQuery {
    allProjects {
      nodes {
        id
        name
        description
        localeStrings {
          langCode
          name
          description
        }
      }
    }
  }
`;
