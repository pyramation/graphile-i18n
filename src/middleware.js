import DataLoader from 'dataloader';
import langParser from 'accept-language-parser';
import { ACCEPTED_LANGUAGES } from './env';

const escapeIdentifier = (str) => `"${str.replace(/"/g, '""')}"`;

export const makeLanguageDataLoaderForTable = (_req) => {
  const cache = new Map();
  return (
    props,
    pgClient,
    languageCodes,
    identifer,
    idType,
    sqlField,
    gqlField
  ) => {
    let dataLoader = cache.get(props);
    if (!dataLoader) {
      const { table, coalescedFields, variationsTableName, key } = props;
      const schemaName = escapeIdentifier(table.namespaceName);
      const baseTable = escapeIdentifier(table.name);
      const variationTable = escapeIdentifier(variationsTableName);
      const joinKey = escapeIdentifier(key);
      const fields = coalescedFields.join(', ');
      const b = [schemaName, baseTable].join('.');
      const v = [schemaName, variationTable].join('.');
      dataLoader = new DataLoader(async (ids) => {
        const { rows } = await pgClient.query(
          `
            select *
            from unnest($1::${idType}[]) ids(${identifer})
            inner join lateral (
              select b.${identifer}, v.${sqlField} as "${gqlField}", ${fields}
              from ${b} b
              left join ${v} v
              on (v.${joinKey} = b.${identifer} and array_position($2, ${sqlField}) is not null)
              where b.${identifer} = ids.${identifer}
              order by array_position($2, ${sqlField}) asc nulls last
              limit 1
            ) tmp on (true)
            `,
          [ids, languageCodes]
        );
        return ids.map((id) => rows.find((r) => r.id === id));
      });
      cache.set(props, dataLoader);
    }
    return dataLoader;
  };
};

export const additionalGraphQLContextFromRequest = (req, res) => {
  const language = langParser.pick(
    ACCEPTED_LANGUAGES,
    req.get('accept-language')
  );

  // Accept-Language: *
  // Accept-Language: en-US,en;q=0.5
  // Accept-Language: fr-CH, fr;q=0.9, en;q=0.8, de;q=0.7, *;q=0.5
  return {
    langCodes: [language], // in future make fallback languages
    getLanguageDataLoader: makeLanguageDataLoaderForTable(req)
  };
};
