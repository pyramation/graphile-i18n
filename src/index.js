export * from './plugin';
export * from './middleware';

import { LangPlugin } from './plugin';

export default LangPlugin;
