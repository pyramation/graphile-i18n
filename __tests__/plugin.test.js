import '../utils/env';
import { env, snapshot } from 'graphile-test';
import { GraphQLTest } from '../utils/graphile-test';
import { GetProjectsAndLanguages } from '../utils/queries';
import { LangPlugin, makeLanguageDataLoaderForTable } from '../src';

const { SCHEMA } = env;

const getDbString = () =>
  `postgres://${env.PGUSER}:${env.PGPASSWORD}@${env.PGHOST}:${env.PGPORT}/${env.PGDATABASE}`;

const { setup, teardown, graphQL } = GraphQLTest(
  {
    appendPlugins: [LangPlugin],
    schema: SCHEMA,
    graphqlRoute: '/graphql',
    graphileBuildOptions: {
      langPluginLanguageCodeGqlField: 'langCode',
      langPluginLanguageCodeColumn: 'lang_code',
      langPluginAllowedTypes: ['citext', 'text'],
      langPluginDefaultLanguages: ['en']
    }
  },
  {
    langCodes: ['es'],
    getLanguageDataLoader: makeLanguageDataLoaderForTable()
  },
  getDbString()
);

beforeAll(async () => {
  await setup();
});
afterAll(async () => {
  await teardown();
});

it('works', async () => {
  await graphQL(async (query) => {
    const data = await query(GetProjectsAndLanguages);
    expect(snapshot(data)).toMatchSnapshot();
  });
});
