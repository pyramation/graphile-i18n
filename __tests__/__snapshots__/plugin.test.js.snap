// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`works 1`] = `
Object {
  "data": Object {
    "allProjects": Object {
      "nodes": Array [
        Object {
          "description": "project one",
          "id": "[ID]",
          "localeStrings": Object {
            "description": "proyecto uno",
            "langCode": "es",
            "name": "proj 1",
          },
          "name": "Proj1",
        },
        Object {
          "description": "project two",
          "id": "[ID]",
          "localeStrings": Object {
            "description": "proyecto dos",
            "langCode": "es",
            "name": "proj 2",
          },
          "name": "Proj2",
        },
        Object {
          "description": "project three",
          "id": "[ID]",
          "localeStrings": Object {
            "description": "proyecto tres",
            "langCode": "es",
            "name": "proj 3",
          },
          "name": "Proj3",
        },
      ],
    },
  },
}
`;
